### PR TITLE
[ContactStructuralMechanicsApplication] Migrating MPCContactApplication

### PR DESCRIPTION
**Pull request by loumalouomega**
Migrating to master

### CODE REVIEW

**Review by philbucher — COMMENTED** [review]
:+1: 
what is the difference btw the existing contact and the one with MPCs?
your contact mpc seems to not do anything

BTW the test files are large, can you reduce the size?
Or maybe reuse the existing test-mdpas?

**Comment by loumalouomega** [comment]
> +1
> what is the difference btw the existing contact and the one with MPCs?

This one is simpler, is basically a "NTN", but does not add LM

> your contact mpc seems to not do anything

For now it is just a linear constraint, but I let it as an independent in case of future expansion

> 
> BTW the test files are large, can you reduce the size?

The files I think already contain the minimum

> Or maybe reuse the existing test-mdpas?

I will try

File: applications/ContactStructuralMechanicsApplication/custom_processes/mpc_contact_search_wrapper_process.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+            size_2 = size_1;
+        }
+
+        // Once both defined we break
+        if (size_1 > 0 && size_2 > 0)
+            break;
+    }
+
+    // Removing to avoid problems
+    ThisParameters.RemoveValue("simple_search");
```
Comment (RiccardoRossi) [inline applications/ContactStructuralMechanicsApplication/custom_processes/mpc_contact_search_wrapper_process.cpp:55]: why don't you just ignore it... it is generally not nice to remove entries from settings (i am aware it is done elsewhere...still it is not nice)

**Review by RiccardoRossi — COMMENTED** [review]
overall it looks good to me, even though it looks a bit scary with 88 files.

@AlejandroCornejo can u check if it works correctly?

**Comment by AlejandroCornejo** [comment]
I'll try ASAP

File: applications/ContactStructuralMechanicsApplication/custom_processes/mpc_contact_search_wrapper_process.cpp (hunk 1 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_processes/mpc_contact_search_wrapper_process.cpp:55]: This is done in the other wrapper, I can remove it in both places in a future PR

**Comment by loumalouomega** [comment]
@RiccardoRossi the number of files is normal, I migrated a whole app with many tests

**Comment by RiccardoRossi** [comment]
ok, i let @AlejandroCornejo to approve

**Comment by AlejandroCornejo** [comment]
Working on it now

**Comment by loumalouomega** [comment]
@RiccardoRossi I think there is a bug in OMP in the BlockBuilderAndSolver, some tests with MPC fail saometimes with 8 threads, I commented the #pragma in the BlockBuilderAndSolver and they don't fail anymore. I don't know exactly where the problem is coming from

**Comment by RiccardoRossi** [comment]
Everything is possible, nevertheless the Builder and solver is quite tested, so my first guess is that the error is somewhere else. 
Which loop did u comment out concretely?
any clue of why we would have an error?

**Comment by loumalouomega** [comment]
> Everything is possible, nevertheless the Builder and solver is quite tested, so my first guess is that the error is somewhere else.

Nope, I have checked, I commented all OMP in my app and I only get rid of the errors commenting the BlockBuilderAndSolver

> Which loop did u comment out concretely?

All of them!!

> any clue of why we would have an error?

Nope, I am testing right now. But I think is related with the reactions when considering constraints (as the formulation considers the reaction for several evaluations)

**Comment by RiccardoRossi** [comment]
The reaction calculation is identical with or without MPCs, so i really don't think it can come out of that

**Comment by loumalouomega** [comment]
> The reaction calculation is identical with or without MPCs, so i really don't think it can come out of that

In any case I am not lying, the errors disappear when commenting the OMP loops in the B&S. I am decommenting one by one (34), but as it is highly used i takes long to compile each time

**Comment by loumalouomega** [comment]
@RiccardoRossi Looks like it is the build :P (yes, the Build)

**Comment by AlejandroCornejo** [comment]
I confirm that the behaviour is the same that the old MPCContactApp. Thank you @loumalouomega

**Review by AlejandroCornejo — APPROVED** [review]
If the others agree, you can merge

**Comment by RiccardoRossi** [comment]
@loumalouomega the build of the system or the Build of the reactions?

what is the problem concretely??

**Comment by loumalouomega** [comment]
> @loumalouomega the build of the system or the Build of the reactions?
> 
> what is the problem concretely??

The global build, **the Build**, I couldn't do more tests because I leave my workplace to do some paperwork. I will come back in 4 hours (I live far from the place, sorry)

**Comment by loumalouomega** [comment]
@RiccardoRossi looks like the system is ill-conditioned and the OMP slightly changes the system (I did not expected this with MPC), can be easily solved changing the solver. It is here #5642

**Pull request merged**